azure-application-insights-spring-boot-starter/src/main/java/com/microsoft/applicationinsights/boot/ApplicationInsightsModuleConfiguration.java

```diff
@@ -0,0 +1,72 @@
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.boot;
+
+import com.microsoft.applicationinsights.TelemetryConfiguration;
+import com.microsoft.applicationinsights.boot.conditional.ConditionalOnOperatingSystem;
+import com.microsoft.applicationinsights.boot.conditional.OperatingSystem;
+import com.microsoft.applicationinsights.boot.initializer.SpringBootContextInitializer;
+import com.microsoft.applicationinsights.extensibility.initializer.DeviceInfoContextInitializer;
+import com.microsoft.applicationinsights.extensibility.initializer.SdkVersionContextInitializer;
+import com.microsoft.applicationinsights.internal.perfcounter.ProcessPerformanceCountersModule;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.env.Environment;
+
+/**
+ * {@link Configuration} for non-web applications.
+ *
+ * @author Arthur Gavlyukovskiy
+ */
+@Configuration
+@ConditionalOnBean(TelemetryConfiguration.class)
+public class ApplicationInsightsModuleConfiguration {
+
+    @Bean
+    public SpringBootContextInitializer springBootContextInitializer(Environment environment) {
+        return new SpringBootContextInitializer(environment);
+    }
+
+    @Bean
+    public SdkVersionContextInitializer sdkVersionContextInitializer() {
+        return new SdkVersionContextInitializer();
+    }
+
+    @Bean
+    public DeviceInfoContextInitializer deviceInfoContextInitializer() {
+        return new DeviceInfoContextInitializer();
+    }
+
+    @Bean
+    @ConditionalOnOperatingSystem(OperatingSystem.WINDOWS)
+    @ConditionalOnProperty(value = "azure.application-insights.default-modules.ProcessPerformanceCountersModule.enabled", havingValue = "true", matchIfMissing = true)
+    public ProcessPerformanceCountersModule processPerformanceCountersModule() {
+        try {
+            return new ProcessPerformanceCountersModule();
+        }
+        catch (Exception e) {
+            throw new IllegalStateException("Could not initialize Windows performance counters module", e);
```

Comment:
I will change this to logging error.

Reply:
Thanks.

Reply:
I remember why I did it, actually you can't just log error here because bean instance must be returned. I'm not quite sure how to fix this properly.

Reply:
Although I don't see any reason why it would throw an exception. `ProcessBuiltInPerformanceCountersFactory` is instance of `WindowsPerformanceCountersFactory` and if something goes wrong inside this factory it returns empty list.

Reply:
Seems like very unlikely situation. Can't you return null in such case?

Reply:
I don't think it's possible, in best case we can either create conditional where we try to create instance and return true only if it was successfully created, or have message with clear action to do, something like 'We are failed to create performance metric, please set `azure.application-insights.default-modules.ProcessPerformanceCountersModule.enabled=false` to avoid this error message'. But in general I don't see any chance for this exception to be thrown in real code. @dhaval24 what do you think?

Reply:
@gavlyukovskiy from technical point of view it's possible, such bean declaration is 100% valid and it works:
`@Bean
	public CustomBean customBean(){
		return null;
	}`

Reply:
You're right, it's valid. But then I should check that module is not null when initializing them in, I can do this as well, but it just seem useless to me to add a lot of checks that verify situation that could not be exist. That's why I throw Illegal**State**Exception here. If there are will be issue with that it indicates bug on our side and you can't save user from all bugs you may ever have. Although user has simple switch to disable this module so it won't be a blocker anyways.

Reply:
@gavlyukovskiy I am fine with failing and asking to set the message  please set azure.application-insights.default-modules.ProcessPerformanceCountersModule.enabled=false to avoid this error message'. Please add this conditional.